mlir-tensorrt/compiler/include/mlir-tensorrt/Compiler/OptionsProviders.h

```diff
@@ -27,13 +27,21 @@
 #include "mlir-tensorrt-dialect/Utils/Options.h"
 #include "mlir/Support/LLVM.h"
 #include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Error.h"
 #include <string>
 
 namespace mlirtrt::compiler {
 
+struct OptionsProvider {
+  virtual ~OptionsProvider() {}
+  virtual void addToOptions(mlir::OptionsContext &context) = 0;
+
+  virtual llvm::Error finalize() { return llvm::Error::success(); }
+};
+
 /// DebugOptions are options that are common to different compiler API
 /// interfaces.
-struct DebugOptions {
+struct DebugOptions : public OptionsProvider {
```

Comment:
I can use CRTP here if we don't want vtable overhead

Reply:
If we didn't need a base class to require `DebugOptions` to have `addToOptions`, then why did we need one to add `finalize`?

Reply:
@christopherbate It's so we can have a default implementation since most of the classes don't actually need to implement it. We could alternatively use some template magic to not call `finalize` at all if it's not defined:
```cpp
if constexpr (hasFinalize<OptionsProviderT>())
    result->finalize();
```
That's not too difficult to do with SFINAE (and obviously we'd factor it out so it doesn't pollute the call-sites).

Reply:
That sounds like a good idea

Reply:
Added